Move catchAll to Math

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46354 Remove catchAllKernel_.
* **#45939 Move catchAll to Math**



Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

Commit:
[WIP] Move catchAll to Math

[ghstack-poisoned]

Commit:
Update on "[WIP] Move catchAll to Math"




[ghstack-poisoned]

Commit:
Update on "[WIP] Move catchAll to Math"




[ghstack-poisoned]

Commit:
Update on "[WIP] Move catchAll to Math"




[ghstack-poisoned]

Commit:
Update on "[WIP] Move catchAll to Math"




[ghstack-poisoned]

Commit:
Update on "[WIP] Move catchAll to Math"


Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move catchAll to Math"


Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move catchAll to Math"


Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move catchAll to Math"


Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move catchAll to Math"


Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move catchAll to Math"


Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move catchAll to Math"


Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move catchAll to Math"


Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move catchAll to Math"


Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move catchAll to Math"


Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move catchAll to Math"


Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

Commit:
Update on "Move catchAll to Math"




Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

Commit:
Update on "Move catchAll to Math"




Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

ailzhang commented:
Context: https://github.com/pytorch/pytorch/issues/46231

Commit:
Update on "Move catchAll to Math"




Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

Commit:
Update on "Move catchAll to Math"




Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

Commit:
Update on "Move catchAll to Math"




Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

Commit:
Update on "Move catchAll to Math"




Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -259,6 +257,8 @@ std::pair<const AnnotatedKernel&, const char*> OperatorEntry::computeDispatchTab
     //      fit 2.1 and we can remove 2.4 entirely.
     if (!has_backend_kernel && !catchAllKernel_.empty()) {
       TORCH_INTERNAL_ASSERT(catchAllKernel_.front().kernel.isValid());
+      // Prepare for catchAll removal, make sure it's not used in dispatchTable
```
I'm not sure the assert above this one (line 260) makes sense anymore...

ailzhang replied:
Yea these two code blocks indeed can be removed, I want to keep them as it is so that if anything went wrong (there's other ways registering to catchAll that I'm not aware of), we can track them easily. 
Both code blocks are removed in the next PR ;)

bhosmer commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -272,6 +272,8 @@ std::pair<const AnnotatedKernel&, const char*> OperatorEntry::computeDispatchTab
   // 4. Catch all
   if (!catchAllKernel_.empty()) {
     TORCH_INTERNAL_ASSERT(catchAllKernel_.front().kernel.isValid());
+    // Prepare for catchAll removal, make sure it's not used in dispatchTable
```
same for the assert above here (line 274)

bhosmer commented on aten/src/ATen/core/op_registration/op_registration_test.cpp:
```diff
@@ -789,8 +789,8 @@ TEST(OperatorRegistrationTest, whenRegisteringBackendFallbackKernelAndCatchallKe
 
   called = false;
   auto stack = callOp(*op, dummyTensor(c10::DispatchKey::CPU), "hello ");
-  EXPECT_FALSE(called);
```
Does it make sense to move the modified tests over to NewOperatorRegistrationTest?

ailzhang replied:
Yep! I'll move them in this PR!

bhosmer reviewed (approved):
LG!

Commit:
Update on "Move catchAll to Math"




Differential Revision: [D24165890](https://our.internmc.facebook.com/intern/diff/D24165890)

[ghstack-poisoned]
